### ISSUE TITLE
Building --disable-shared --enable-static results in being unable to link MPI apps

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#9986]
Configuring Open MPI master HEAD with `--disable-shared --enable-static` results in being unable to link MPI apps:

```sh
$ git rev-parse HEAD
acbe7b0ca1cf48afa0e7acc2760bde8a8172e7a8
$ git submodule status
 680331773926b62c245626dbc9cf78aed2d641d3 3rd-party/openpmix (v1.1.3-3327-g68033177)
 78825642e8594ebffda0942fa04e375077819732 3rd-party/prrte (psrvr-v2.0.0rc1-4147-g78825642e8)
```

```sh
$ rm -rf $HOME/bogus
$ ./configure --prefix=$HOME/bogus --enable-static --disable-shared |& tee config.out
...
$ make -j 32 install |& tee make.out
...
$ cd examples
$ make clean
...
$ mpicc -g  hello_c.c  -o hello_c
/home/jsquyres/bogus/lib/libopen-pal.a(reachable_netlink_utils_common.o): In function `opal_reachable_netlink_rt_raw_parse_cb':
reachable_netlink_utils_common.c:(.text+0xd): undefined reference to `nlmsg_hdr'
reachable_netlink_utils_common.c:(.text+0x44): undefined reference to `nlmsg_data'
reachable_netlink_utils_common.c:(.text+0x64): undefined reference to `nlmsg_parse'
reachable_netlink_utils_common.c:(.text+0x77): undefined reference to `nla_get_u32'
reachable_netlink_utils_common.c:(.text+0xa4): undefined reference to `nlmsg_data'
reachable_netlink_utils_common.c:(.text+0xb0): undefined reference to `nlmsg_size'
reachable_netlink_utils_common.c:(.text+0x10e): undefined reference to `nla_get_u32'
/home/jsquyres/bogus/lib/libopen-pal.a(reachable_netlink_utils_common.o): In function `opal_reachable_netlink_rt_lookup':
reachable_netlink_utils_common.c:(.text+0x188): undefined reference to `nl_socket_alloc'
reachable_netlink_utils_common.c:(.text+0x19e): undefined reference to `nl_connect'
reachable_netlink_utils_common.c:(.text+0x1ae): undefined reference to `nl_socket_disable_seq_check'
reachable_netlink_utils_common.c:(.text+0x1c3): undefined reference to `nl_socket_get_fd'
reachable_netlink_utils_common.c:(.text+0x222): undefined reference to `nlmsg_alloc_simple'
reachable_netlink_utils_common.c:(.text+0x245): undefined reference to `nlmsg_append'
reachable_netlink_utils_common.c:(.text+0x255): undefined reference to `nla_put_u32'
reachable_netlink_utils_common.c:(.text+0x265): undefined reference to `nla_put_u32'
reachable_netlink_utils_common.c:(.text+0x26d): undefined reference to `nlmsg_hdr'
reachable_netlink_utils_common.c:(.text+0x278): undefined reference to `nl_socket_get_local_port'
reachable_netlink_utils_common.c:(.text+0x295): undefined reference to `nlmsg_set_proto'
reachable_netlink_utils_common.c:(.text+0x2ab): undefined reference to `nl_send'
reachable_netlink_utils_common.c:(.text+0x2b6): undefined reference to `nlmsg_free'
reachable_netlink_utils_common.c:(.text+0x2f6): undefined reference to `nl_socket_modify_cb'
reachable_netlink_utils_common.c:(.text+0x305): undefined reference to `nl_recvmsgs_default'
reachable_netlink_utils_common.c:(.text+0x334): undefined reference to `nl_close'
reachable_netlink_utils_common.c:(.text+0x33c): undefined reference to `nl_socket_free'
reachable_netlink_utils_common.c:(.text+0x364): undefined reference to `nl_close'
reachable_netlink_utils_common.c:(.text+0x36c): undefined reference to `nl_socket_free'
reachable_netlink_utils_common.c:(.text+0x389): undefined reference to `nl_geterror'
reachable_netlink_utils_common.c:(.text+0x3a9): undefined reference to `nl_geterror'
reachable_netlink_utils_common.c:(.text+0x3c2): undefined reference to `nl_socket_free'
reachable_netlink_utils_common.c:(.text+0x42a): undefined reference to `nl_geterror'
reachable_netlink_utils_common.c:(.text+0x451): undefined reference to `nl_geterror'
reachable_netlink_utils_common.c:(.text+0x493): undefined reference to `nl_geterror'
/home/jsquyres/bogus/lib/libpmix.a(compress_zlib.o): In function `zlib_decompress':
compress_zlib.c:(.text+0x8e): undefined reference to `inflateInit_'
compress_zlib.c:(.text+0xb8): undefined reference to `inflate'
compress_zlib.c:(.text+0xc2): undefined reference to `inflateEnd'
/home/jsquyres/bogus/lib/libpmix.a(compress_zlib.o): In function `zlib_compress.part.0':
compress_zlib.c:(.text+0x166): undefined reference to `deflateInit_'
compress_zlib.c:(.text+0x171): undefined reference to `deflateBound'
compress_zlib.c:(.text+0x1b1): undefined reference to `deflate'
compress_zlib.c:(.text+0x1bc): undefined reference to `deflateEnd'
compress_zlib.c:(.text+0x256): undefined reference to `deflateEnd'
/home/jsquyres/bogus/lib/libpmix.a(compress_zlib.o): In function `decompress_string':
compress_zlib.c:(.text+0x376): undefined reference to `inflateInit_'
compress_zlib.c:(.text+0x3a1): undefined reference to `inflate'
compress_zlib.c:(.text+0x3ab): undefined reference to `inflateEnd'
collect2: error: ld returned 1 exit status
```

If we manually add the missing libraries, it compiles fine, of course:

```sh
$ mpicc -g hello_c.c -o hello_c -lz -lnl-3
$ 
```

pkg-config does exist.  If I manually set `PKG_CONFIG_PATH`:

```sh
$ export PKG_CONFIG_PATH=$bogus/lib/pkgconfig
$ pkg-config pmix --cflags
-I/home/jsquyres/bogus/include 
$ pkg-config pmix --libs
-L/home/jsquyres/bogus/lib -lpmix 
$ pkg-config pmix --libs --static
-L/home/jsquyres/bogus/lib -lpmix -Wl,-rpath -Wl,/home/jsquyres/bogus/lib -lz 
```

PMIx reported libz.  Looking at the output from configure, it looks like the libnl3 dependency is coming from PMIx, too (and also OMPI's OPAL).  FWIW:

```sh
$ pkg-config ompi-c --libs
-L/home/jsquyres/bogus/lib -Wl,-rpath -Wl,/home/jsquyres/bogus/lib -Wl,-rpath -Wl,/home/jsquyres/bogus/lib -Wl,-rpath -Wl,/home/jsquyres/bogus/lib -Wl,--enable-new-dtags -lmpi 
$ pkg-config ompi-c --libs --static
-L/home/jsquyres/bogus/lib -Wl,-rpath -Wl,/home/jsquyres/bogus/lib -Wl,-rpath -Wl,/home/jsquyres/bogus/lib -Wl,-rpath -Wl,/home/jsquyres/bogus/lib -Wl,--enable-new-dtags -lmpi -lopen-pal -ldl -levent_core -levent_pthreads -lhwloc -lpmix -lpthread -lrt -lm -lutil
```

**Comment by awlauria** [comment]
I tried this on rhel8.4 and am unable to reproduce

```
./configure --prefix=`pwd`/exports --enable-static --disable-shared --with-pmix=internal --with-prrte=internal --disable-man-pages --with-hwloc=internal
```

```
$ pkg-config --static --libs ./exports/lib/pkgconfig/pmix.pc
-L/home/awl/ompi_build/exports/lib -lpmix -Wl,-rpath -Wl,/home/awl/ompi_build/exports/lib -lz -levent_core -levent_pthreads 
```

```
$ pkg-config --static --libs ./exports/lib/pkgconfig/ompi.pc
-L/home/awl/ompi_build/exports/lib -Wl,-rpath -Wl,/home/awl/exports/lib -Wl,-rpath -Wl,/home/awl/ompi_build/exports/lib -Wl,-rpath -Wl,/home/awl/ompi_build/exports/lib -Wl,--enable-new-dtags -lmpi -lopen-pal -ldl -levent_pthreads -levent_core -lhwloc -lpmix -lpthread -lrt -lm -lutil 
```

**Comment by bwbarrett** [comment]
@jsquyres can you attach the output of configure and the resulting config.log?

**Comment by jsquyres** [comment]
[config.log](https://github.com/open-mpi/ompi/files/8026668/config.log)
[config.stdout.txt](https://github.com/open-mpi/ompi/files/8026669/config.stdout.txt)

**Comment by awlauria** [comment]
Your pkg-config didn't pick up libevent for some reason. Not sure if that is what caused this, but it is strange.

**Comment by jsquyres** [comment]
This has been fixed on both main and v5.0.x.  Closing.
